### PR TITLE
SequenceCache: Sequence change protection

### PR DESCRIPTION
**Pull request by servantftransperfect**
[wip] Fix a crash

### CODE REVIEW

File: src/qtAliceVision/SequenceCache.cpp [hunk 1]
```diff
@@ -260,55 +270,79 @@ void SequenceCache::onPrefetchingProgressed(int)
...
 
-void SequenceCache::onPrefetchingDone(int reqFrame)
+void SequenceCache::onPrefetchingDone(int sequenceId, int reqFrame)
 {
-    // Update internal state
-    _loading = false;
+    //Make sure the fetching concerns the actual sequence Id
+    bool exitOld = false;
+    _lockSequence.lock();
+    if (sequenceId != sequenceId)
```
Comment (mugulmd) [inline src/qtAliceVision/SequenceCache.cpp]: I think there might be a typo: shouldn't it be `sequenceId != _sequenceId` instead ?

File: src/qtAliceVision/SequenceCache.hpp [hunk 2]
```diff
@@ -144,6 +148,15 @@ class SequenceCache : public QObject, public ImageServer {
     /// Target size used to compute downscale
     int _targetSize;
 
+    /// Local threadpool
+    QThreadPool _threadPool;
```
Comment (mugulmd) [inline src/qtAliceVision/SequenceCache.hpp:152]: why not use the global thread pool instance ?

File: src/qtAliceVision/SequenceCache.cpp [hunk 3]
```diff
@@ -260,55 +270,79 @@ void SequenceCache::onPrefetchingProgressed(int)
...
+    {   
+        exitOld = true;
+    }
+    if (reqFrame >= _sequence.size())
```
Comment (mugulmd) [inline src/qtAliceVision/SequenceCache.cpp:282]: if the previous check passed then reqFrame should be in the correct range so this check is redundant (or maybe I missed something ?)

File: src/qtAliceVision/SequenceCache.cpp (hunk 1 above)
Comment (servantftransperfect) [reply src/qtAliceVision/SequenceCache.cpp]: yes. wouhou :/

File: src/qtAliceVision/SequenceCache.hpp (hunk 2 above)
Comment (servantftransperfect) [reply src/qtAliceVision/SequenceCache.hpp:152]: using the global thread pool lead to crashes sometimes. I do not have more explanations than this !

File: src/qtAliceVision/SequenceCache.cpp (hunk 3 above)
Comment (servantftransperfect) [reply src/qtAliceVision/SequenceCache.cpp:282]: ceinture et bretelle :-)

**Review by mugulmd — APPROVED** [review]

**Review by cbentejac — APPROVED** [review]

**Pull request merged**